contracts/Strategy.sol

```diff
@@ -53,80 +140,224 @@ contract Strategy is BaseStrategy {
         returns (uint256 _liquidatedAmount, uint256 _loss)
     {
-        // TODO: Do stuff here to free up to `_amountNeeded` from all positions back into `want`
-        // NOTE: Maintain invariant `want.balanceOf(this) >= _liquidatedAmount`
-        // NOTE: Maintain invariant `_liquidatedAmount + _loss <= _amountNeeded`
-
         uint256 totalAssets = want.balanceOf(address(this));
         if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+            uint256 amountToFree = _amountNeeded.sub(totalAssets);
+
+            (, uint256 lastAccrued, ) = kashiPair.accrueInfo();
+            if (block.timestamp > lastAccrued) {
+                // We need to call accrue to accurately calculate totalAssets
+                kashiPair.accrue();
+            }
+
+            uint256 deposited = estimatedTotalAssets().sub(totalAssets);
+
+            if (deposited < amountToFree) {
+                amountToFree = deposited;
+            }
+            if (deposited > 0) {
+                uint256 sharesToFree =
+                    bentoBox.toShare(
+                        BIERC20(address(want)),
+                        amountToFree,
+                        true
+                    );
+                uint256 sharesToFreeFromKashi =
+                    sharesToFree.sub(
+                        bentoBox.balanceOf(
+                            BIERC20(address(want)),
+                            address(this)
+                        )
+                    );
+
+                if (sharesToFreeFromKashi > 0) {
+                    kashiPair.removeAsset(
+                        address(this),
+                        _toKashiFraction(sharesToFreeFromKashi, true)
+                    );
+                }
+
+                bentoBox.withdraw(
+                    BIERC20(address(want)),
+                    address(this),
+                    address(this),
+                    0,
+                    bentoBox.balanceOf(BIERC20(address(want)), address(this))
+                );
+            }
+
+            _liquidatedAmount = want.balanceOf(address(this));
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
     function liquidateAllPositions() internal override returns (uint256) {
-        // TODO: Liquidate all positions and return the amount freed.
+        liquidatePosition(type(uint256).max);
         return want.balanceOf(address(this));
     }
 
-    // NOTE: Can override `tendTrigger` and `harvestTrigger` if necessary
-
+    // The _newStrategy must support the same kashiPair or bad things will happen
     function prepareMigration(address _newStrategy) internal override {
-        // TODO: Transfer any non-`want` tokens to the new strategy
-        // NOTE: `migrate` will automatically forward all `want` in this strategy to the new one
+        kashiPair.transfer(_newStrategy, kashiPair.balanceOf(address(this)));
```

Comment:
shouldn't this withdraw from kashi lending instead?

Reply:
Nvm,that isnt needed here

Reply:
Keeping everything in bentobox should save gas